internal/provider/resource_application.go

```diff
@@ -164,6 +165,34 @@ func (r *applicationResource) Schema(_ context.Context, _ resource.SchemaRequest
 					stringplanmodifier.UseStateForUnknown(),
 				},
 			},
+			"storage": schema.SetNestedAttribute{
+				Description: "Configure storage constraints for the juju application.",
+				Optional:    true,
+				NestedObject: schema.NestedAttributeObject{
+					Attributes: map[string]schema.Attribute{
+						"label": schema.StringAttribute{
+							Description: "The specific storage option defined in the charm.",
+							Required:    true,
+						},
+						"size": schema.StringAttribute{
+							Description: "The size of each volume. E.g. 100G.",
+							Required:    true,
+						},
+						"pool": schema.StringAttribute{
+							Description: "Name of the storage pool to use. E.g. ebs on aws.",
+							Optional:    true,
+						},
+						"count": schema.Int64Attribute{
+							Description: "The number of volumes.",
+							Optional:    true,
+						},
+					},
+					PlanModifiers: []planmodifier.Object{
+						objectplanmodifier.RequiresReplaceIfConfigured(),
+						objectplanmodifier.UseStateForUnknown(),
+					},
```

Comment:
On reflection, we may want to make these plan modifiers on each attribute, rather than on the object themselves.  I need to understand the difference better.